File: backend/tools/tool_domain_search.py
"""
Инструмент «Поиск по домену» — использует модуль modules/domain_search.py.
WHOIS, DNS, SSL, crt.sh без внешнего UI (Flet не используется).
"""
import os
import sys
from typing import Any

from .base import ToolBase

# Корень проекта (родитель backend/) для импорта modules
_ROOT = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
if _ROOT not in sys.path:
    sys.path.insert(0, _ROOT)


def _normalize_domain(domain: str) -> str:
    s = (domain or "").strip().lower()
    for prefix in ("https://", "http://"):
        if s.startswith(prefix):
            s = s[len(prefix) :]
            break
    if "/" in s:
        s = s.split("/")[0]
    if s.startswith("www."):
        s = s[4:]
    return s


class ToolDomainSearch(ToolBase):
    @property
    def tool_id(self) -> str:
        return "domain-search"

    @property
    def name(self) -> str:
        return "Поиск по домену"

    @property
    def description(self) -> str:
        return "Информация по домену: WHOIS, DNS, SSL-сертификат, поддомены (crt.sh)."

    def run(self, params: dict[str, Any]) -> dict[str, Any]:
        domain_raw = (params.get("domain") or "").strip()
        domain = _normalize_domain(domain_raw)
        if not domain:
            return {"success": False, "error": "Укажите домен (например example.com)."}
        if " " in domain:
            return {"success": False, "error": "Некорректный домен."}
        if "." not in domain:
            return {"success": False, "error": "Введите домен с точкой (example.com)."}

        try:
            from modules.domain_search import run_domain_search
            result = run_domain_search(domain)
        except ImportError as e:
            return {"success": False, "error": "Модуль domain_search не найден: " + str(e)[:100]}
        except Exception as e:
            return {"success": False, "error": str(e)[:300]}

        if not result.get("success"):
            return {"success": False, "error": result.get("error", "Ошибка модуля"), "output": result.get("output", "")}
        return {"success": True, "output": result.get("output", "")}

File: static/js/tools/tool-domain-search.js
/**
 * Панель «Поиск по домену» — модуль modules/domain_search (WHOIS, DNS, SSL, crt.sh).
 */
import { apiRun } from "../app.js";

const TOOL_ID = "domain-search";

function normalizeDomain(s) {
  if (!s || typeof s !== "string") return "";
  let t = s.trim().toLowerCase();
  if (t.startsWith("https://")) t = t.slice(8);
  else if (t.startsWith("http://")) t = t.slice(7);
  const slash = t.indexOf("/");
  if (slash !== -1) t = t.slice(0, slash);
  if (t.startsWith("www.")) t = t.slice(4);
  return t;
}

export function initPanel(container) {
  container.innerHTML = `
    <div class="tool-view tool-domain-search-page">
      <section class="tool-view-section tool-view-section--search">
        <h2 class="tool-view-section__title">ПОИСК ПО ДОМЕНУ</h2>
        <p class="tool-cli-desc">Введите домен — получите WHOIS, DNS, SSL-сертификат и поддомены с crt.sh. Домен можно ввести с https:// — будет приведён к виду example.com.</p>
        <form id="form-domain-search" class="tool-cli-form">
          <div class="form-group">
            <label for="ds-domain">Домен</label>
            <input type="text" id="ds-domain" name="domain" placeholder="example.com или https://example.com" autocomplete="off">
          </div>
          <div class="tool-form-actions">
            <button type="submit" class="btn btn-primary" id="btn-ds-run">Поиск</button>
            <button type="button" class="btn btn-secondary" id="btn-ds-clear">Очистить</button>
          </div>
        </form>
        <div id="ds-form-error" class="form-error" role="alert"></div>
      </section>
      <section class="tool-view-section tool-view-section--terminal">
        <h2 class="tool-view-section__title">РЕЗУЛЬТАТ</h2>
        <div id="ds-status" class="tool-cli-status" role="status"></div>
        <pre id="ds-output" class="tool-terminal"></pre>
      </section>
    </div>
  `;

  const form = container.querySelector("#form-domain-search");
  const formErrorEl = container.querySelector("#ds-form-error");
  const statusEl = container.querySelector("#ds-status");
  const outputEl = container.querySelector("#ds-output");
  const btn = container.querySelector("#btn-ds-run");
  const btnClear = container.querySelector("#btn-ds-clear");

  function setFormError(msg) {
    formErrorEl.textContent = msg || "";
    formErrorEl.style.display = msg ? "block" : "none";
  }

  if (btnClear) btnClear.addEventListener("click", () => {
    form.querySelector("#ds-domain").value = "";
    outputEl.textContent = "";
    outputEl.className = "tool-terminal";
    statusEl.textContent = "";
    statusEl.className = "tool-cli-status";
    setFormError("");
  });

  form.addEventListener("submit", async (e) => {
    e.preventDefault();
    const domain = normalizeDomain(form.querySelector("#ds-domain").value);
    setFormError("");
    if (!domain) {
      setFormError("Укажите домен (например example.com).");
      return;
    }
    if (domain.includes(" ")) {
      setFormError("Домен не должен содержать пробелы.");
      return;
    }

    const btnText = btn.textContent;
    btn.textContent = "Подождите…";
    btn.disabled = true;
    btn.classList.add("is-loading");
    statusEl.textContent = "Запрос WHOIS, DNS, SSL, crt.sh…";
    statusEl.className = "tool-cli-status";
    outputEl.textContent = "";
    outputEl.style.display = "block";

    try {
      const data = await apiRun(TOOL_ID, { domain });
      statusEl.textContent = data.success ? "Готово." : (data.error || "Ошибка.");
      if (!data.success) statusEl.className = "tool-cli-status error";
      const outRaw = (data.output || data.error || "").trim();
      outputEl.textContent = outRaw || (data.error ? "" : "Нет данных по домену.");
      outputEl.className = "tool-terminal " + (data.success ? "success" : "error");
    } catch (err) {
      statusEl.className = "tool-cli-status error";
      statusEl.textContent = err.message || "Ошибка запроса.";
      outputEl.textContent = err.message || "Ошибка сети";
      outputEl.className = "tool-terminal error";
    }
    btn.textContent = btnText;
    btn.disabled = false;
    btn.classList.remove("is-loading");
  });
}
